run.sh

```diff
@@ -97,30 +96,51 @@ exec_to_load="$1"
 shift
 
 arguments="-m 2G -nographic -nodefaults "
-arguments+="-display none -serial stdio -device isa-debug-exit "
-arguments+="-fsdev local,security_model=passthrough,id=hvirtio0,path=$rootfs_9p "
-arguments+="-device virtio-9p-pci,fsdev=hvirtio0,mount_tag=fs0 "
-arguments+="-kernel $kvm_image "
-arguments+="-initrd $exec_to_load "
+arguments="${arguments}-display none -serial stdio -device isa-debug-exit "
+arguments="${arguments}-fsdev local,security_model=passthrough,id=hvirtio0,path=$rootfs_9p "
+arguments="${arguments}-device virtio-9p-pci,fsdev=hvirtio0,mount_tag=fs0 "
+arguments="${arguments}-kernel $kvm_image "
+arguments="${arguments}-initrd $exec_to_load "
 
 if test "$use_kvm" -eq 1; then
-    arguments+="-enable-kvm -cpu host "
+  file="/dev/kvm"
+  if ! [ -c "$file" ]; then
+    echo "$file does not exist, cannot run with KVM" 2>&1
+    exit 1
+  fi
+
+  # Check if the effective user has RW permission to /dev/kvm,
+  # if they haven't, temporarily add the user to the group that owns
+  # /dev/kvm, by default the KVM group. If that also fails, run QEMU
+  # as root, but with the -runas parameter to drop privileges.
+  if ! [ -r "$file" ] || ! [ -w "$file" ]; then
+    group_owner=$(stat -c "%G" $file)
+    if [ "$(stat -c "%A" $file | cut -c 5-6)" = "rw" ] && ! [ "$group_owner" = "$(id -un 0)" ];
```

Comment:
I would actually look if the current user is part of that group. The fact that the group is `root` is not that relevant. What would matter is if the user is part of the group. Probably something like:

```bash
... && [ groups | grep -q "$group_owner" ]
```

Reply:
If the user is part of the group, and the group has `rw` permissions on `/dev/kvm`, then the check on line 116 will succeed and line 118 is not evaluated.

Line 118 is for the case in which a user is not part of the group that owns `/dev/kvm`. In that case, it will temporarily add the user to that group if:
- that group does have `rw` permissions, in case `/dev/kvm` is `chmod 600`.
- that group isn't root, in case `/dev/kvm/` is `chown root:root`.

In the first case, temporarily adding the user to a group doesn't allow you to r/w the file and you need root privileges. In the second case, you preferably want to drop the root group privileges before qemu starting guest execution. qemu only has the `-runas` parameter to drop user root privileges, but this can only be passed when qemu is run as root.

In both these cases, you 'fallback' to the old behavior of running qemu as root, but with the `-runas` parameter.

Reply:
Got it, thanks.